File: C# Fundamentals/C# OOP Advanced/07.Iterators_Comparators/Ex/P01_ListyIterator/ListyIterator.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class ListyIterator<T>
{
    private List<T> elements;

    private int index;

    public ListyIterator(params T[] elements)
    {
        this.elements = new List<T>(elements);
    }

    public void Print()
    {
        if (this.elements.Count == 0)
        {
            throw new InvalidOperationException("Invalid Operation!");
        }
        Console.WriteLine(this.elements[this.index]);
    }
    
    public bool Move()
    {
        if(this.index + 1 < this.elements.Count)
        {
            this.index++;
            return true;
        }
        return false;
    }
    public bool HasNext()
    {
        if(this.index + 1 > this.elements.Count - 1)
        {
            return false;
        }
        return true;
    }
}

File: Java/01.Advanced/01.Advanced/17.IteratorsComparators/Ex/src/t_01_listy_iterator/ListyIterator.java
package t_01_listy_iterator;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class ListyIterator implements Iterator<String> {
    private int index;

    private List<String> collection;

    public ListyIterator(String... collection) {
        this.collection = List.of(collection);
    }

    public boolean move() {
        if (hasNext()) {
            index++;

            return true;
        }
        return false;
    }

    public void print() {
        if (collection == null || collection.size() < 1 || index == collection.size()) {
            System.out.println("Invalid Operation!");
        } else {
            System.out.println(collection.get(index));
        }
    }

    @Override
    public boolean hasNext() {
        return index < collection.size() - 1;
    }

    @Override
    public String next() {
        return null;
    }
}
